[v1.13] CI IPsec backports

qmonnet opened the pull request:
Here's a bunch of IPsec CI backports labelled as `backport/author`, before the workflows diverge too much.

Should be similar to [the v1.14 equivalent](https://github.com/cilium/cilium/pull/29966), although I needed to backport two preparatory commits to make `kind.sh` behave.

 * [x] #25317 (@chancez)
 * [ ] #26338 (@ysksuzuki)
 * [x] #29485 (@brb)
 * [x] #29514 (@brb)
 * [x] #29793 (@qmonnet)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
for pr in 25317 26338 29485 29514 29793; do contrib/backporting/set-labels.py $pr done 1.13; done
```
or with
```
make add-labels BRANCH=v1.13 ISSUES=25317,26338,29485,29514,29793
```

chancez committed:
Set SED binary from Makefile in kind.sh

[ upstream commit cb2b1124e57d9c66da20af7875066509f4746d75 ]

Our Makefiles and scripts largely assume GNU sed, and kind.sh uses `sed
-z` which requires GNU sed. MacOS ships with BSD sed, so to make the
existing `kind.sh` script work, make the sed binary configurable, and
pass through the Makefile SED to the script.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

ysksuzuki committed:
contrib/kind: Add secondary iface to KIND network

[ upstream commit d675d6fd7d3c185bdff9c1943cfae31e3714f8c9 ]

Add secondary iface to KIND network, so that we can test the NodePort
via secondary iface.

[ Backport note: Minor conflict on kind-down.sh. ]

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

brb committed:
kind.sh: Allow to specify apiServer addr + port

[ upstream commit c9296a5b3ce477d7bcab393b844339a471e84ff7 ]

This is going to be used when exposing a K8s cluster in a LVH VM to its
hypervisor host ("0.0.0.0:6443").

[ Backport note:
    Minor conflicts due to the script supporting more arguments on main
    branch ]

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

brb committed:
gh/workflows: Add lvh-kind action

[ upstream commit b6255d061aeea88db51b3ccc6861bfc40188825a ]

The action provisions K8s cluster on LVH VM, and then exposes it to GHA
runner.

With this action we will be able to run kubectl / cilium / etc commands
directly from GHA runner instead of SSH'ing into VM.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

brb committed:
ci-e2e: Use lvh-kind action instead of c/little-vm-helper

[ upstream commit 936c8ae01f87de2d9155d80a0f1c8bce0cd48fc1 ]

[ Backport notes:
    - Solved conflicts in conformance-e2e.yaml, because of
      cilium/little-vm-helper version that was upgraded on the branch
    - Replaced "cilium-dbg" from main to simply "cilium"
    - Ignored the step to debug issue 26062, not present on this branch
    - Don't wait for spire-server/spire-agent to be deployed
    - Adjust the way we pass EXTRA to take into account recent backports
      on the branch ]

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

brb committed:
ci-ipsec-e2e: Use lvh-kind

[ upstream commit f8e61f3c740c4d90783cde7954487b89cee0458f ]

[ Backport notes:
    - Minor conflict on "Setup Conn Disrupt Test"'s command:
      --conn-disrupt-dispatch-interval passed to cilium-cli
    - Minor conflict on "Perform Conn Disrupt Test"'s command:
      disabling node-to-node-encryption check passed to cilium-cli
    - Minor conflicts in conformance-ipsec-e2e.yaml: Using "cilium"
      instead of "cilium-dbg"
    - Ignore the step to debug issue 26062, not present on this branch
    - Use v1.13 version of lvh-kind instead of main ]

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

brb committed:
ci-ipsec-upgrade: Use lvh-kind

[ upstream commit cd93d3774959645dc3f12157c2544a3d9c16df57 ]

Reduces the GH workflow complexity, and will allow us to reuse the
workflow for non-Kind based tests.

[ Backport notes:
    - Minor conflict: "IMAGE=quay.io/cilium/kindest-node:${k8s_version}"
      not present in this branch (the block is removed in this commit
      anyway)
    - Minor conflict: Keep --conn-disrupt-dispatch-interval 0ms
    - Minor conflict: Remove note on the need to update cilium to
      cilium-dbg
    - Ignore the step to debug issue 26062, not present on this branch
    - Use v1.13 version of lvh-kind instead of main ]

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

brb committed:
ci-e2e: Use lvh-kind in secure way

[ upstream commit 654d92f29c4f47a4e98d87b14aca03166707493e ]

Otherwise, an external PR can inject malicious cmds in the action.

[ Backport note: Use v1.13 version of lvh-kind instead of main ]

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
ci/ipsec: Skip waiting for images when skipping upgrade/dowgrade

[ upstream commit 6ae935476e296986bd9e8ed7d3335748a3f3f727 ]

The workflow to test upgrades/downgrades with IPsec supports two cases:
upgrading from the last minor version, or from the last patch release.
In the case of patch releases, there are instances where we are not able
to determine the latest patch releases. This is the case on the current
development branch, where no patch releases have been created yet, for
example. In such a case, we skip most of the steps in the job.

When implementing the check on the patch release value and the bypass of
the relevant steps, we omitted to skip the step where we wait for the CI
images to be available, potentially causing unnecessary wait time for
the job. Let's skip this step as well. Since we're at it, we can also
skip the step for setting Kind parameters.

Fixes: c9dedb49f5a6 ("ci/ipsec: Skip upgrade/downgrade test to patch release on main branch")
Fixes: cd93d3774959 ("ci-ipsec-upgrade: Use lvh-kind")
Reported-by: Paul Chaignon <paul.chaignon@gmail.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

jschwinger233 reviewed (approved):
brb's commits look good, thanks!

qmonnet committed:
ci-ipsec-upgrade: Update name of test to skip, to match cilium-cli's

Due to flakiness, we skip a test in workflow tests-ipsec-upgrade.yaml by
using the relevant cilium-cli flag. However, the name of this test in
cilium-cli has changed over time [0]. We recently pulled that change
when bumping the cilium-cli from v0.15.14 to v0.15.19, but we haven't
updated the name of the test in the workflow. Let's do it now.

[0] https://github.com/cilium/cilium-cli/commit/4880c91a726ddbad3363cc48f4bb8c11ed1150bd

Fixes: 2c2aa5079f17 ("chore(deps): update dependency cilium/cilium-cli to v0.15.19")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
ci-e2e: Bump cilium-cli, skip node-to-node-encryption with tunneling

In a later backport commit, we will rework some CI workflows, including
conformance-e2e.yaml. One consequence is that we will unintentionally
"fix" an issue introduced in a recent backport for this workflow, but
this will cause cilium-cli to complain and the workflow to fail. This is
a preparatory commit containing three adjustments to fix the issue.

The issue is that, since commit ee6c73a07cb4 ("ci-e2e: Do not check for
missed tail calls with host-fw"), we sometimes pass the flag
--expected-drop-reasons with the variable ${EXTRA[@]} to cilium-cli. But
this does not work as expected, because:

- The flag was only added in cilium-cli v0.15.18 (the workflow uses
  v0.15.14). This should cause cilium-cli to complain.

- This complaint was masked by another issue: when passing ${EXTRA[@]}
  as a cmd to Action cilium/little-vm-helper, the dollar symbol should
  be escaped to avoid expansing the variable before it was initialised
  and eventually passing an empty string instead of the expected
  arguments. So currently, we never pass the flag (and never cause
  cilium-cli to complain).

So in this commit, we make sure that \${EXTRA[@]} is passed properly,
and we bump cilium-cli (to v0.15.19) to ensure the flag
--expected-drop-reasons is available.

But the cilium-cli version bump has yet another consequence: in the
connectivity tests, v0.15.14 used to skip the node-to-node-encryption
test if node encryption was disabled. This is no longer the case with
v0.15.19 [0]. So cilium-cli attempts to run node-to-node-encryption, but
fails with tunneling is enabled (the test is broken). This is the third
adjustment brought by this commit: we address this by skipping
node-to-node-encryption manually in the workflow when tunneling is
enabled.

[0] https://github.com/cilium/cilium-cli/commit/32e12a8e5ab25c54ffe5366825e420a30ea4c3fe

Fixes: ee6c73a07cb4 ("ci-e2e: Do not check for missed tail calls with host-fw")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno reviewed (approved):
Approving for the two new commits. Thanks Quentin!

pchaigno commented on .github/workflows/tests-ipsec-upgrade.yaml:
```diff
@@ -349,11 +333,10 @@ jobs:
-          extra-connectivity-test-flags: --test '!no-missed-tail-calls'
+          # Disable no-unexpected-packet-drops (formerly no-missed-tail-calls)
+          # due to https://github.com/cilium/cilium/issues/26739
+          extra-connectivity-test-flags: --test '!no-unexpected-packet-drops'
```
That doesn't look correct actually. We were skipping only Missed tail calls and now we'll skip all packet drops. We should use `--expected-drop-reasons` instead. No sure where my head was at.

qmonnet replied:
@pchaigno Ouch, yeah I see what you mean. My bad. Are you working on a fix? If not, I can do it today (and for https://github.com/cilium/cilium/pull/30211).

pchaigno replied:
I'm not. Go for it :slightly_smiling_face:

rgo3 replied:
I just found this PR while looking at how to add `--expected-drop-reasons "+Failed to update or lookup TC buffer"`. Should `!no-missed-tail-calls` also come back?

pchaigno replied:
Nop, `no-missed-tail-calls` doesn't exist anymore (unless I messed up).

rgo3 replied:
yeah I didn't ask the right question. I meant if I should also add missed tail calls as an expected drop reason? I assume we still need to skip those failures.

qmonnet replied:
(I suppose we need `--expected-drop-reasons=+Missed tail call`?)
[EDIT - Sorry, my page didn't refresh and I didn't see the rest of this conversation below my earlier message, until now :facepalm:]

pchaigno replied:
Ah, yes. We need both Missed tail calls and CT buffer drops to be allowlisted.

qmonnet replied:
I see Robin has this covered in https://github.com/cilium/cilium/pull/30249
